marklogic-data-hub/src/main/resources/ml-modules/root/data-hub/4/triggers/entity-model-validate-trigger.xqy

```diff
@@ -0,0 +1,86 @@
+xquery version '1.0-ml';
+
+import module namespace es = "http://marklogic.com/entity-services"
+  at "/MarkLogic/entity-services/entity-services.xqy";
+import module namespace hent = "http://marklogic.com/data-hub/hub-entities" at "/data-hub/5/impl/hub-entities.xqy";
+import module namespace tde = "http://marklogic.com/xdmp/tde"
+        at "/MarkLogic/tde.xqy";
+import module namespace trgr = 'http://marklogic.com/xdmp/triggers' at '/MarkLogic/triggers.xqy';
+
+declare variable $ENTITY-MODEL-COLLECTION as xs:string := "http://marklogic.com/entity-services/models";
+
+declare variable $trgr:uri as xs:string external;
+
+let $entity-def := fn:doc($trgr:uri)/object-node()
+(: Create map version so we can update $ref to external links if we have to :)
+let $entity-def-map as map:map := $entity-def
+let $entity-base-uri := $entity-def/info/baseUri
+let $entity-title := $entity-def/info/title
+let $entity-version := $entity-def/info/version
+let $validate-base-uri := try {
+    xdmp:json-validate-node($entity-def/info, xdmp:unquote('{type: "object", properties: {"baseUri":{type:"string", format:"uri"}}, required: ["baseUri"]}'), "strict")
+  } catch ($ex) {
+    if(($ex/error:code eq "XDMP-JSVALIDATEINVFORMAT" or $ex/error:code eq "XDMP-JSVALIDATEMISSING") and fn:contains($ex/error:format-string/text(), "baseUri")) then (
+      fn:error(fn:QName("error","XDMP-JSVALIDATEINVFORMAT"),fn:concat("A valid baseUri is required (ex: http://www.example.org/) for entity: ", $entity-title, "."), fn:concat("Error Message: ", $ex/error:format-string/text()))
```

Comment:
I tested this out on a new project with a single entity model file with an invalid baseUri. When I install the app using QS, the deployment fails because of the error thrown by the entity validation trigger, though it happens late enough in the process that QS still reports back to the user that the installation succeeds. 

The entity shows up in the UI though - QS still loads the file. But the entity document isn't in staging or final (nor are any flows or steps). 

If I edit the entity, there isn't any indication that there's an error with the Base URI. If I save it, I get the error message on Base URI. 

So while this prevents a user from creating an entity with a bad Base URI, I agree with @bsrikan's comment about this being a breaking change, and it seems like it'll be difficult for a user to figure out that the breaking change occurs if they're using QS.

Reply:
The UI just doesn't handle errors.

It's not a 'breaking change' anymore than anytime a deploy fails silently in QS.

I'd imagine the UI to report failed save attempts to be its own ticket, as this doesn't modify any of that behavior.

At a certain point we have to decide on a) allowing users to insert bad entities like we do now, which causes failed tdes and models which breaks then during flow running or b) force compliance.

It's not a breaking change anymore than it used to be. We get an issue a day about invalid models silently not being deployed.

Reply:
That's fair. I think it's worth making a note of in our release notes though - i.e. that if you have entity models without a base URI, you'll need to add one. What's the best way to ensure something is added to the release notes? Pinging @sbayatpur and @eltesoro-ml